TextBreakIteratorCache shouldn't be used from worker threads.

mattwoodrow opened the pull request:
#### 3e4dfc234c15f9f98e8398e10dae9945bc160981
<pre>
TextBreakIteratorCache shouldn&apos;t be used from worker threads.
<a href="https://bugs.webkit.org/show_bug.cgi?id=260498">https://bugs.webkit.org/show_bug.cgi?id=260498</a>
&lt;rdar://113654067&gt;

Reviewed by David Kilzer.

An alternative here would be to have a cache per-thread, but I&apos;m not aware of an easy way to
store thread local data from WTF currently.

This just skips the cache for non-main threads, and we can consider adding it back as needed
for performance in the future.

* Source/WTF/wtf/text/TextBreakIterator.h:
(WTF::CachedTextBreakIterator::CachedTextBreakIterator):
(WTF::CachedTextBreakIterator::~CachedTextBreakIterator):

Canonical link: <a href="https://commits.webkit.org/267175@main">https://commits.webkit.org/267175@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f060c65de4f7d997d25bc4734d4f0c13b400b8c3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/15861 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/16180 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/16561 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/17624 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/14887 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/19190 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/16278 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/17357 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/16052 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/16524 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/13513 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/18381 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/13773 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/14330 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/21213 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/13638 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/14765 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/14495 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/17746 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/15102 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/15089 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/12791 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/16060 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/14330 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/4024 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3788 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/18699 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/17224 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/14911 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/3815 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

ddkilzer reviewed (approved):
r=me

ddkilzer commented on Source/WTF/wtf/text/TextBreakIterator.h:
```diff
@@ -181,13 +182,13 @@ class CachedTextBreakIterator {
     WTF_MAKE_FAST_ALLOCATED;
 public:
     CachedTextBreakIterator(StringView string, const UChar* priorContext, unsigned priorContextLength, TextBreakIterator::Mode mode, const AtomString& locale, TextBreakIterator::ContentAnalysis contentAnalysis = TextBreakIterator::ContentAnalysis::Mechanical)
-        : m_backing(TextBreakIteratorCache::singleton().take(string, priorContext, priorContextLength, mode, contentAnalysis, locale))
```
Should we add Debug assertions in `TextBreakIteratorCache::singleton().take()` and `put()` that they are always called on the main thread?  (This doesn't block landing, but it would catch issues quicker in the future.)

smfr commented on Source/WTF/wtf/text/TextBreakIterator.h:
```diff
@@ -181,13 +182,13 @@ class CachedTextBreakIterator {
 public:
     CachedTextBreakIterator(StringView string, const UChar* priorContext, unsigned priorContextLength, TextBreakIterator::Mode mode, const AtomString& locale, TextBreakIterator::ContentAnalysis contentAnalysis = TextBreakIterator::ContentAnalysis::Mechanical)
-        : m_backing(TextBreakIteratorCache::singleton().take(string, priorContext, priorContextLength, mode, contentAnalysis, locale))
+        : m_backing(isMainThread() ? TextBreakIteratorCache::singleton().take(string, priorContext, priorContextLength, mode, contentAnalysis, locale) : TextBreakIterator(string, priorContext, priorContextLength, mode, contentAnalysis, locale))
```
Don't we want `isMainRunLoop`?

mattwoodrow replied:
I believe `isMainRunLoop` is for WebKit code, `isMainThread` is for WebCore and WTF.

mattwoodrow committed:
TextBreakIteratorCache shouldn't be used from worker threads.
https://bugs.webkit.org/show_bug.cgi?id=260498
<rdar://113654067>

Reviewed by David Kilzer.

An alternative here would be to have a cache per-thread, but I'm not aware of an easy way to
store thread local data from WTF currently.

This just skips the cache for non-main threads, and we can consider adding it back as needed
for performance in the future.

* Source/WTF/wtf/text/TextBreakIterator.h:
(WTF::CachedTextBreakIterator::CachedTextBreakIterator):
(WTF::CachedTextBreakIterator::~CachedTextBreakIterator):

Canonical link: https://commits.webkit.org/267175@main
